### PR TITLE
`Programming exercises`: Add C# programming language template

### PR DESCRIPTION
**Pull request by magaupp**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a local server configured with **LocalVC** and **Jenkins**.

### Description

This template adds C# using .NET 8 and NUnit. Usage is demonstrated using the usual Strategy Pattern example.

### Steps for Testing

Prerequisites:
- 1 Course

1. Navigate to the exercise creation page
2. Select `C#` as the programming language
3. Fill out the rest of the required fields. Keep the default build plan
4. Generate the exercise
5. Wait until the `Template` and `Solution` build jobs finish
6. The `Template Result` should show 0 out of 10 passed tests
7. The `Solution Result` should show 16 out of 16 passed tests

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review
#### Manual Tests
- [x] Test

## Release Notes

- **New Features**
	- Added support for the C# programming language, including integration into the plagiarism detection service and continuous integration systems.
	- Introduced new project templates and scripts for C# exercises, including sorting algorithms and build configurations.
	- Updated documentation to reflect C# support and its features.

- **Bug Fixes**
	- Enhanced error handling and coverage verification for programming languages.

- **Tests**
	- Implemented comprehensive unit tests for new C# functionalities, including sorting algorithms and structural integrity checks.

- **Chores**
	- Updated project configurations to utilize the latest .NET SDK and Gradle versions.

### CODE REVIEW

File: src/main/resources/templates/c_sharp/solution/ISortStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+namespace assignment;
+
+public interface ISortStrategy
+{
+    /// <summary>
+    /// Sorts a list of Dates.
+    /// </summary>
+    /// <param name="dates">the List of Dates to be sorted</param>
+    void PerformSort(List<DateTime> dates);
+}
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/solution/ISortStrategy.cs:10]: This is left as a bonus exercise

File: src/main/resources/templates/c_sharp/exercise/BubbleSort.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+namespace assignment;
+
+public class BubbleSort
+{
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/exercise/BubbleSort.cs:4]: Files under the `exercise` directory are incomplete exercises to be completed by the student.

File: src/main/resources/templates/c_sharp/exercise/BubbleSort.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    /// <summary>
+    /// Sorts dates with BubbleSort.
+    /// </summary>
+    /// <param name="dates">the List of Dates to be sorted</param>
+    public void PerformSort(List<DateTime> dates)
+    {
+        //TODO: implement
+    }
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/exercise/BubbleSort.cs:12]: Files under the `exercise` directory are incomplete exercises to be completed by the student.

File: src/main/resources/templates/c_sharp/exercise/MergeSort.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+namespace assignment;
+
+public class MergeSort
+{
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/exercise/MergeSort.cs:4]: Files under the `exercise` directory are incomplete exercises to be completed by the student.

File: src/main/resources/templates/c_sharp/exercise/MergeSort.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    /// <summary>
+    /// Sorts dates with MergeSort.
+    /// </summary>
+    /// <param name="dates">the List of Dates to be sorted</param>
+    public void PerformSort(List<DateTime> input)
+    {
+        // TODO: implement
+    }
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/exercise/MergeSort.cs:12]: Files under the `exercise` directory are incomplete exercises to be completed by the student.

File: src/main/resources/templates/c_sharp/solution/Policy.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    private const int DATES_SIZE_THRESHOLD = 10;
+
+    private Context context;
+
+    public Policy(Context context)
+    {
+        this.context = context;
+    }
+
+    public Context Context { get => context; set => context = value; }
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/solution/Policy.cs:14]: We use a non-nullable type here

File: src/main/resources/templates/c_sharp/exercise/.gitignore [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+# Local History for Visual Studio Code
+.history/
+
+# Windows Installer files from build outputs
+*.cab
+*.msi
+*.msix
+*.msm
+*.msp
+
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/exercise/.gitignore:402]: This is the default file provided by the `dotnet new gitignore` command.

File: src/main/resources/templates/c_sharp/test/.gitignore [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+
+# Local History for Visual Studio Code
+.history/
+
+# Windows Installer files from build outputs
+*.cab
+*.msi
+*.msix
+*.msm
+*.msp
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/test/.gitignore:401]: This is the default file provided by the dotnet new gitignore command.

File: src/main/resources/templates/c_sharp/test/StructuralTest.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
+using System.Reflection;
+
+namespace test;
+
```
[the remark this answers is not in the document]
Comment (magaupp) [reply src/main/resources/templates/c_sharp/test/StructuralTest.cs:4]: This is in the global usings.

**Review by HawKhiem — APPROVED** [review]
Tested on TS2. Works as described

image

**Review by kevinfischer4 — APPROVED** [review]
Tested on TS2. Works as described.
Bildschirmfoto 2024-11-01 um 12 23 15

**Review by ItsaaaMeMario — APPROVED** [review]
Tested on TS1, everything worked as expected. The template result showed 0 out of 16 passed tests and the solution result showed 16 out of 16 passed tests.

**Review by Cathy0123456789 — APPROVED** [review]
Tested on TS1. Works as described.

**Review by sawys777 — APPROVED** [review]
Tested on TS1, created an exercise named "Testing for Csharp" and everything worked as described.

**Review by zagemello — APPROVED** [review]
Tested on TS2, all worked as expected

**Review by BBesrour — APPROVED** [review]
code lgtm

**Pull request merged**